tests/Mqtt5ClientTest.cpp

```diff
@@ -1936,6 +1939,234 @@ static int s_TestMqtt5WillTest(Aws::Crt::Allocator *allocator, void *)
 }
 AWS_TEST_CASE(Mqtt5WillTest, s_TestMqtt5WillTest)
 
+/*
+ *  Shared Subscription test
+ */
+static int s_TestMqtt5SharedSubscriptionTest(Aws::Crt::Allocator *allocator, void *)
+{
+    Mqtt5TestEnvVars mqtt5TestVars(allocator, MQTT5CONNECT_IOT_CORE);
+    if (!mqtt5TestVars)
+    {
+        printf("Environment Variables are not set for the test, skip the test");
+        return AWS_OP_SKIP;
+    }
+
+    ApiHandle apiHandle(allocator);
+
+    String currentUUID = Aws::Crt::UUID().ToString();
+
+    const String TEST_TOPIC = "test/MQTT5_Binding_CPP_" + currentUUID;
+    const String sharedTopicFilter = "$share/crttest/test/MQTT5_Binding_CPP_" + currentUUID;
+
+    const int MESSAGE_NUMBER = 10;
+    std::atomic<int> client_messages(0);
+    bool client1_received = false;
+    bool client2_received = false;
+
+    std::vector<int> receivedMessages;
+    for (int i = 0; i < MESSAGE_NUMBER; i++)
+    {
+        receivedMessages.push_back(0);
+    }
+
+    Aws::Iot::Mqtt5ClientBuilder *subscribe_builder =
+        Aws::Iot::Mqtt5ClientBuilder::NewMqtt5ClientBuilderWithMtlsFromPath(
+            mqtt5TestVars.m_hostname_string,
+            mqtt5TestVars.m_certificate_path_string.c_str(),
+            mqtt5TestVars.m_private_key_path_string.c_str(),
+            allocator);
+    ASSERT_TRUE(subscribe_builder);
+
+    std::promise<void> client_received;
+    auto onMessage_client1 = [&](const PublishReceivedEventData &eventData) -> int {
+        String topic = eventData.publishPacket->getTopic();
+        if (topic == TEST_TOPIC)
+        {
+            ByteCursor payload = eventData.publishPacket->getPayload();
+            String message_string = String((const char *)payload.ptr, payload.len);
+            int message_int = atoi(message_string.c_str());
+            ASSERT_TRUE(message_int < MESSAGE_NUMBER);
+            ++receivedMessages[message_int];
+            client1_received = true;
+
+            bool exchanged = false;
+            int desired = 11;
+            int tested = 10;
+            client_messages++;
+            exchanged = client_messages.compare_exchange_strong(tested, desired);
```

Comment:
Trivial: Callbacks will be called on a single thread (the eventloop thread), so no data race should be possible. It's not a big deal, but I'd simplify it for maintainability.
I think, it's safe to make `client_messages` non-atomic.

Reply:
Nope, I'm wrong, so nevermind

Reply:
as discussed off line, threads might come on different threads as well...